### ISSUE TITLE
Implement simple viewer

### ISSUE THREAD

**Issue by MrNeRF** [MrNeRF/gaussian-splatting-cuda#64]
- Should run in its own gui thread.
- It should reuse the rasterizer
- Should allow for simple controls like fps or orbiting.

**Comment by panxkun** [comment]
hello, I'd be happy to take on this task.

The following video demo is from [liteviz-gs](https://github.com/panxkun/liteviz-gs), a lightweight C++/OpenGL-based visualization project. Compared to the original SIBR-based visualizer in the 3DGS repository, it has fewer dependencies and is much easier to build, so I believe it can be integrated into this project quite smoothly.

Before I get started, I’d like to clarify a few points regarding the requirements:

Is online (real-time) visualization needed, i.e., visualizing the training process as it runs?

https://github.com/user-attachments/assets/5d54acea-c024-46e1-bf98-61e15b78f92c

**Comment by MrNeRF** [comment]
Hey @panxkun ,
that would be great! Ideally we could preview and control the training by hitting the start and stop button and change the parameters before starting the training.

So yes, a live view would be very appreciated.

I was considering to pull in your visualizer anyway so that sounds like a perfect opportunity.

**Comment by MrNeRF** [comment]
It would be also great if we could just load and view a splat without training of course.

**Comment by panxkun** [comment]
Sure, it sounds like I can start by introducing a simple solution for loading and offline visualization to get familiar with the code, and then move on to implementing a more complete online training visualization process, that sounds great.

**Comment by MrNeRF** [comment]
Sounds like a plan!

main.cpp outlines the structure which is very simple. The trainer.cpp/hpp defines the training process. splat_data.hpp/cpp owns the relevant data which is used for training but can be visualized as well.

Rasterizer.cpp outlines the forward pass. If you want to save an image, image_io.hpp has some basic interface. That's about it.

**Comment by panxkun** [comment]
I've already added the basic visualization part to the code, but there are a few issues I'm not entirely sure about, so I'd like to discuss them.

Following the original 3DGS visualization logic from SIBR, I retrieve images from `gs::rasterizer` during the training loop based on camera viewpoints obtained via mouse interaction, and display them in a Viewer window. However, this slightly slows down the training process. The problem with this approach is that by defining the Viewer inside the Trainer and inserting the visualization logic there, it becomes difficult to implement the originally intended feature of interactively modifying training parameters, since all training-related configurations are completed before training starts. Therefore, I believe that a basic version that simply visualizes the training process is acceptable for now.

In addition, I noticed that when the window resolution is not a multiple of 8, pixel misalignment occurs. I suspect this might be due to a resolution requirement in the gsplat backend, although this is just my hypothesis.

Functionally speaking, most of the implementation has been completed, but the remaining unresolved and potential issues may still take considerable time to address.

**Comment by MrNeRF** [comment]
Ideally, there would be a “Start Training” button that loads the parameters from the UI before initiating training. The GUI should be optionally enabled, right? That way, we can train without visualization. A --viz flag would be useful for that.

I agree, we don't need a full scale implementation at this stage. These kinds of features can be added later.

The visualizer could send an update request every second or so and trigger a render when the camera changes. I also think the visualizer should have a handle to the trainer so training can be started, stopped, or resumed (later implemented so not in scope here). The Viewer just needs access to splat_data via a pointer. That data access should be protected with a lock.

Of course, this introduces some slowdown, but it would only occur when visualization is enabled.

> In addition, I noticed that when the window resolution is not a multiple of 8, pixel misalignment occurs. I suspect this might be due to a resolution requirement in the gsplat backend, although this is just my hypothesis.

Unfortunately, at this point I don't yet have an intuition about what’s causing this issue.

> Functionally speaking, most of the implementation has been completed, but the remaining unresolved and potential issues may still take considerable time to address.

Great, understood. Thank you. Hope that helps?

**Comment by panxkun** [comment]
Hello, I’ve now organized most of the code and believe it’s ready as a basic version. However, I still have a crucial unresolved issue, so I’d like to raise it here to see if anyone can offer insights.

This video shows the visualization of the training process, which satisfies the first three initial requirements. For the next stage, I also plan to add configuration support for the initial training parameters.

https://github.com/user-attachments/assets/595dd332-9d14-486e-b8ff-c2ae1a032f52

Here is the current issue: as previously mentioned, when I resize the window, the entire image becomes misaligned. Intuitively, it seems like a memory alignment issue. However, this only happens when the window is stretched horizontally—vertical resizing works fine. Unfortunately, I haven't found any clues or entry points to investigate the root cause.

https://github.com/user-attachments/assets/9d3d503c-7f8f-4e68-a80b-468d3220e2e4
